src/survey.ts

```diff
@@ -0,0 +1,212 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import { commands, MessageItem, Uri, window, workspace } from 'vscode';
+import { callWithTelemetryAndErrorHandling, IActionContext } from "vscode-azureextensionui";
+import { configPrefix } from './constants';
+import { ext } from "./extensionVariables";
+import { assert } from './fixed_assert';
+import { httpGet } from "./httpGet";
+import { hoursToMs, minutesToMs, weeksToMs } from "./util/time";
+
+namespace stateKeys {
+    export const neverShowSurvey = 'neverShowSurvey';
+    export const surveyPostponedUntilTime = 'surveyPostponedUntilTime';
+}
+
+const linkToSurvey = 'https://aka.ms/arm-tools-survey';
+const surveyPrompt = "Could you please take 2 minutes to tell us how well the Azure Resource Manager Tools extension is working for you?";
+
+// Wait for this amount of time since extension first used (this session) before
+// considering asking the survey (only if still interacting with extension)
+interface ISurveyConstants {
+    activeUsageTimeBeforeSurvey: number;
+    percentageOfUsersToSurvey: number;
+    timeToPostponeAfterYes: number;
+    timeToPostponeAfterLater: number;
+    timeToPostponeAfterNotSelected: number;
+    timeToPostponeAfterNotAccessible: number;
+}
+
+const defaultSurveyConstants: ISurveyConstants = {
+    activeUsageTimeBeforeSurvey: hoursToMs(1),
+    percentageOfUsersToSurvey: 0.25,
+    timeToPostponeAfterYes: weeksToMs(12),
+    timeToPostponeAfterLater: weeksToMs(1),
+    timeToPostponeAfterNotSelected: weeksToMs(12),
+    timeToPostponeAfterNotAccessible: hoursToMs(1)
+};
+const debugSurveyConstants: ISurveyConstants = {
+    activeUsageTimeBeforeSurvey: minutesToMs(1),
+    percentageOfUsersToSurvey: 0.5,
+    timeToPostponeAfterYes: minutesToMs(2),
+    timeToPostponeAfterLater: minutesToMs(1),
+    timeToPostponeAfterNotSelected: minutesToMs(0.5),
+    timeToPostponeAfterNotAccessible: minutesToMs(0.25)
+};
+
+let surveyConstants: ISurveyConstants = defaultSurveyConstants;
+
+// Time user started interacting with extension this session
+let usageSessionStart: number | undefined;
+
+let isReentrant = false;
+let surveyDisabled = false;
+let isDebugMode = false;
+
+export namespace survey {
+    /**
+     * Called whenever the user is interacting with the extension (thus gets called a lot)
+     */
+    export function registerActiveUse(): void {
+        // Don't wait
+        // tslint:disable-next-line: no-floating-promises
+        callWithTelemetryAndErrorHandling("considerShowingSurvey", async (context: IActionContext) => {
+            context.errorHandling.suppressDisplay = true;
+            // This gets called a lot, we don't want telemetry for most of it
+            context.telemetry.suppressIfSuccessful = true;
+
+            if (isReentrant) {
```

Comment:
>isReentrant [](start = 16, length = 11)

Do we need to lock before checking / setting this? If the user opens multiple documents, can this be called multiple times on different threads?

Reply:
JavaScript is single-threaded.  Much easier. :-)